### ISSUE TITLE
Cancel timers when the VisibilityDetector is removed.

### ISSUE THREAD

**Issue by ekarayel** [google/flutter.widgets#12]
# VisibilityDetector

The `VisibilityDetector` should cancel timers, when it is removed from the widget tree.

For example the following test case should work: 
```
void main() {
  testWidgets('Does not crash without waiting for the update interval.',
      (WidgetTester tester) async {
    final _uniqueKey = UniqueKey();
    await tester.pumpWidget(MaterialApp(home: VisibilityDetector(
        key: _uniqueKey,
        child: Container(),
        onVisibilityChanged: (_) {})));
    await tester.pumpWidget(Placeholder());
    // await widgetTester.pump(VisibilityDetectorController.instance.updateInterval);
  }, timeout: Timeout.none);
}
```
But it fails with `Timer still pending`.

Currently we have to uncomment the last line to make it work. This looks like an issue with cancelling timers, or flutter not correctly identifying cancelled timers.

**Comment by jamesderlin** [comment]
In general, if a visible `VisibilityDetector` is removed from a widget tree, then it's transitioning from a visible to a non-visible state and should fire the callback.  The question then becomes: should the callback fire immediately?  If the callback fires immediately, then that leads to a potentially odd situation where some callbacks are fired immediately and others are fired after a delay.  One example would be asymmetry that widgets would report becoming visible after a delay but would report becoming hidden immediately.

However, I think the only reason you care about canceling timers on removal is because of the widget test assertion.  In the case of widget tests, I'd *love* to cancel timers when the widget tree is destroyed in a widget test.  Unfortunately, I'm not aware of a good way to do that.

`VisibilityDetector` detects that it's become hidden when its corresponding `VisibilityDetectorLayer` object has been detached from the `Layer` tree.  But I don't know how to distinguish between becoming detached because it's no longer visible and becoming detached because a widget test is tearing down the widget tree.  This is specific to widget tests (because the widget test framework specifically checks for pending timers and because in the general case, as I noted above, the widgets are becoming non-visible and should fire visibility callbacks).  I consider it to be a bug with the widget test framework that there is no way to clean up the timers in this case (or that the assertion even exists), and that's https://github.com/flutter/flutter/issues/24166.

**Comment by lukasschneiderapps** [comment]
You can use `VisibilityDetectorController.instance.forget(key)` in `dispose()` to cancel the timers of theVisibilityDetector widget

**Comment by juanmendez** [comment]
in `setUp()` or before `await tester.pumpWidget(..)` you can append the following and will resolve the problem
`VisibilityDetectorController.instance.updateInterval = Duration.zero;`

**Comment by jamesderlin** [comment]
Right. ([That's also mentioned in the README file](https://pub.dev/packages/visibility_detector#widget-tests); I'm not sure why I forgot to mention it in the comments above.)
